Sources/BraintreeCore/BTAPIClient.swift

```diff
@@ -50,21 +51,22 @@ import Foundation
         case .tokenizationKey:
             do {
                 self.authorization =  try TokenizationKey(authorization)
-                http = BTHTTP(authorization: self.authorization)
             } catch {
                 return nil
             }
         case .clientToken:
             do {
                 let clientToken = try BTClientToken(clientToken: authorization)
                 self.authorization = clientToken
-
-                http = BTHTTP(authorization: self.authorization)
             } catch {
                 return nil
             }
         }
         
+        let btHttp = BTHTTP(authorization: self.authorization)
+        http = btHttp
+        configurationLoader = ConfigurationLoader(http: btHttp)
```

Comment:
I am thinking about the fact that ConfigurationCache is a singleton and thought of ConfigurationLoader as being one as well. In a weird scenario that there are multiple instances of BTAPIClient. wdyt?

Reply:
This is something we want to do on Android too. Making `ConfigurationLoader` a singleton makes the most sense for our architecture as well.

Reply:
Something that happened in MXO iOS was that almost everything was a singleton. At least in iOS, we need to define the architecture. I think most of the code was a migration from Objective-C to Swift (IMO there are several areas of opportunity that we could refactor.). For now, I think we can proceed this way and discuss whether it would be necessary to convert it to a singleton in the future. Like any design pattern, it has its pros and cons. I'm not entirely a fan because it violates the SRP (Single Responsibility Principle), and we have to think of a creative way to mock

Reply:
@richherrera I'm 100% onboard with that concern. The one thing I noticed with MXO though is that the interface was static e.g. `ConfigurationLoader.load()` . If we go for `ConfigurationLoader.shared.load()` we'll preserve the flexibility needed for mocking if we inject the singleton reference in the constructor (we do this quite a bit in Android with [getInstance()](https://github.com/braintree/braintree_android/blob/586cea65db35fd15013e1317c014e97566325aee/BraintreeCore/src/main/java/com/braintreepayments/api/AnalyticsClient.kt#L27) accessors).

Reply:
Yes, passing it in with default initializer makes testing possible.

Reply:
Necessary, not sure, but I think there might be some weirdness in possibly having multiple instances of BTAPIClient and multiple instances of Loader.

Reply:
@sestevens yeah DIP rules, i like that, something I’d like to know is how we could use the same instance of BTHTTP in both BTAPIClient and ConfigurationLoader. Are there two different instances in Android (for BraintreeClient and ConfigurationLoader)?

Reply:
I think for as long as there can be multiple instances of BTAPIClient, we should address those issues. 

I guess with singleton, you'd need to keep track of BTHTTP instance creation in ConfigurationLoader by authorization or something and check with the getConfiguration function.

I think maybe considering multiple instances of BTAPIClient and configurationLoader, we might need to synchronize ConfigurationCache functions as well some time. 

Great work on this Rich.

Reply:
> Are there two different instances in Android (for BraintreeClient and ConfigurationLoader)?

@richherrera On Android we have a single `BraintreeHttpClient` instance that's shared between `BraintreeClient` and `ConfigurationLoader`. That would be equivalent to `BTAPIClient` and `ConfigurationLoader` sharing a `BTHTTP` instance on iOS.

Reply:
I think that's how we want to go eventually.